File: DSA problems in c++/Quick Sort.cpp
#include<iostream>
using namespace std;

int partition(int arr[], int s, int e)
{
    int pivot = arr[s];

    int cnt = 0;
    for(int i = s+1; i<= e; i++)
    {
        if(arr[i] <= pivot)
            cnt++;

    }

    //place pivot at right place
    int pivotindex = s + cnt;
    swap(arr[pivotindex], arr[s]);

    //handle the left and right part
    int i = s, j = e;

    while(i < pivotindex && j > pivotindex)
    {
        while(arr[i] < pivot)
            i++;
        
        while(arr[j] > pivot)
            j--;
        
        if(i < pivotindex && j > pivotindex)
            swap(arr[i++], arr[j--]);

    }

    return pivotindex;

}

void quicksort(int arr[], int s, int e)
{
    //base case
    if(s >= e)
        return;
    
    //partition
    int p = partition(arr, s, e);

    //sort left part
    quicksort(arr, s, p-1);

    //sort right part
    quicksort(arr, p+1, e);
}

int main()
{
    int arr[5] = {2, 4, 6, 1, 9};
    int n = 5;

    quicksort(arr, 0, n-1);

    for(int i= 0; i< n; i++)
    {
        cout << arr[i] <<" ";
    }
    cout << endl;
    return 0;
}

File: Program's_Contributed_By_Contributors/Rust_Programs/quick_sort.rs
fn quick_sort<T: Ord>(mut arr: Vec<T>) -> Vec<T> {
    if arr.len() <= 1 {
        return arr;
    }

    let pivot = arr.remove(0);
    let mut left = vec![];
    let mut right = vec![];

    for item in arr {
        if item <= pivot {
            left.push(item);
        } else {
            right.push(item);
        }
    }

    let mut sorted_left = quick_sort(left);
    let mut sorted_right = quick_sort(right);

    sorted_left.push(pivot);
    sorted_left.append(&mut sorted_right);

    sorted_left
}

fn main() {
    let arr = vec![10, 80, 30, 90, 40, 50, 70];

    println!("{:?}", arr);
    println!("{:?}", quick_sort(arr));
}
